File: c++/day6.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <set>

int main() {

    std::fstream myFile;
    const int signalLength = 14;

    myFile.open("../inputs/input6.txt", std::ios::in);
    std::string line;
    
    while (std::getline(myFile, line)) {
        int start = 0, end = signalLength;
        for (auto i = 0; i < line.length(); i++) {
            std::string subString = line.substr(start, signalLength);
            std::set<char> signalSet(std::begin(subString), std::end(subString));
            
            if (signalSet.size() == signalLength) {
                std::cout << end << std::endl;
                break;
            }
            start += 1;
            end += 1;
        }
    }
    myFile.close();
}

File: python/day6.py
with open("inputs/input6.txt") as f:
    signal = f.read()
    SIGNAL_LENGTH = 14  # 4 for part 1
    
    end_index = SIGNAL_LENGTH
    
    for i in range(len(signal)):
        sub_string = set(signal[i:end_index])

        if len(sub_string) == SIGNAL_LENGTH:
            print(end_index)
            break
        
        end_index += 1